torch/distributed/_composable/fsdp/_fsdp_param.py

```diff
@@ -120,20 +124,38 @@ def __init__(
         mesh_info: FSDPMeshInfo,
         post_forward_mesh_info: Optional[FSDPMeshInfo],
         device: torch.device,
+        mp_policy: MixedPrecisionPolicy,
     ):
         self._module_info: ParamModuleInfo = module_info
         self.mesh_info = mesh_info
         self.post_forward_mesh_info = post_forward_mesh_info
         self.device = device
-        self._init_dtype_attrs(param)
+        self._init_dtype_attrs(param, mp_policy)
         self._init_sharded_param(param, device)
         if self.post_forward_mesh_info:
             self._init_sharded_post_forward_param_metadata(param)
         self.all_gather_output = torch.empty(0)
         self._param_fqn: Optional[str] = None  # prefixed from root module
 
-    def _init_dtype_attrs(self, param: nn.Parameter):
+    def _init_dtype_attrs(self, param: nn.Parameter, mp_policy: MixedPrecisionPolicy):
+        param_dtype, reduce_dtype = (mp_policy.param_dtype, mp_policy.reduce_dtype)
         self.orig_dtype = param.dtype
+        # Each saved dtype attribute should only be not `None` if it affects
+        # behavior (e.g. requiring casting); otherwise, clamp to `None`
+        if param_dtype is not None and param_dtype == self.orig_dtype:
+            # E.g. orig=compute=bf16
+            param_dtype = None
+        # By default, gradients are computed in the compute dtype
+        if reduce_dtype is not None and (
+            # E.g. orig=compute=reduce=bf16
+            (param_dtype is None and reduce_dtype == self.orig_dtype)
+            # E.g. orig=fp32, compute=reduce=bf16
+            or (param_dtype is not None and reduce_dtype == param_dtype)
```

Comment:
these checks are not readable lol, can we make these checks as instance methods of the MixedPrecisionPolicy?

Reply:
I simplified it considerably by adding a `__post_init__` to `MixedPrecisionPolicy` to clamp `reduce_dtype` to `None` if it matches `param_dtype` and here in `FSDPParam`, only needing to clamp `param_dtype` to `None` if it matches `orig_dtype`.